sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RegionScopedSessionContainer.java

```diff
@@ -0,0 +1,621 @@
+        }
+    }
+
+    @Override
+    public void setSessionToken(String collectionRid, String collectionFullName, Map<String, String> responseHeaders) {
+        throw new NotImplementedException("setSessionToken(String collectionRid, String collectionFullName, Map<String, String> responseHeaders) not implemented for RegionScopedSessionContainer");
+    }
+
+    private void setSessionToken(RxDocumentServiceRequest request, ResourceId resourceId, String collectionName, String token) {
+        String partitionKeyRangeId;
+        ISessionToken parsedSessionToken;
+
+        String[] tokenParts = StringUtils.split(token, ':');
+        partitionKeyRangeId = tokenParts[0];
+        parsedSessionToken = SessionTokenHelper.parse(tokenParts[1]);
+
+        if (logger.isTraceEnabled()) {
+            logger.trace("UPDATE SESSION token {} {} {}", resourceId.getUniqueDocumentCollectionId(), collectionName, tokenParts[1]);
+        }
+
+        boolean isKnownCollection;
+
+        this.readLock.lock();
+        try {
+            isKnownCollection = collectionName != null &&
+                this.collectionNameToCollectionResourceId.containsKey(collectionName) &&
+                this.collectionResourceIdToCollectionName.containsKey(resourceId.getUniqueDocumentCollectionId()) &&
+                this.collectionNameToCollectionResourceId.get(collectionName) == resourceId.getUniqueDocumentCollectionId() &&
+                this.collectionResourceIdToCollectionName.get(resourceId.getUniqueDocumentCollectionId()).equals(collectionName);
+            if (isKnownCollection) {
+                this.addSessionTokenAndTryRecordEpkInBloomFilter(request, resourceId, partitionKeyRangeId, parsedSessionToken);
+            }
+        } finally {
+            this.readLock.unlock();
+        }
+
+        if (!isKnownCollection) {
+            this.writeLock.lock();
+            try {
+                if (resourceId.getUniqueDocumentCollectionId() != 0) {
+                    this.collectionNameToCollectionResourceId.compute(collectionName, (k, v) -> resourceId.getUniqueDocumentCollectionId());
+                    this.collectionResourceIdToCollectionName.compute(resourceId.getUniqueDocumentCollectionId(), (k, v) -> collectionName);
+                }
+                this.addSessionTokenAndTryRecordEpkInBloomFilter(request, resourceId, partitionKeyRangeId, parsedSessionToken);
+            } finally {
+                this.writeLock.unlock();
+            }
+        }
+    }
+
+    private void recordPartitionKeyInBloomFilter(
+        RxDocumentServiceRequest request,
+        Long collectionRid,
+        String regionRoutedTo,
+        PartitionKeyInternal partitionKeyInternal,
+        PartitionKeyDefinition partitionKeyDefinition) {
+
+        if (Strings.isNullOrEmpty(this.firstPreferredReadableRegionCached.get())) {
+            this.firstPreferredReadableRegionCached.set(extractFirstEffectivePreferredReadableRegion());
```

Comment:
curious: why not set it in the constructor? because SDK has not read the database account yet?

Reply:
Correct - although delaying the construction the `RegionScopedSessionContainer` to after resolution of the `DatabaseAccount` in the client-initialization flow should help but don't feel the need to change it either way.